KEP-3000: Image Promotion and Distribution Policy (#3079)
kubernetes/enhancements: keps/sig-release/3000-artifact-distribution/README.md

    - [Process](#process-1)
- [Alternatives / Background](#alternatives--background)
  - [How much is this going to save us?](#how-much-is-this-going-to-save-us)
- [Infrastructure Needed](#infrastructure-needed)
- [Hack on this doc](#hack-on-this-doc)
<!-- /toc -->

## Summary

The container images and release binaries produced by our community need a clear path to be hosted by multiple service/cloud providers.

The global community should be routed to the appropriate mirror for their country or cloud provider to ensure cost effective worldwide access.

This KEP should cover the policy and distribution mechanisms we will put in place to allow creating a globally distributed, multi-cloud and country solution.

## Motivation

Currently we push to a single provider, and distributing to the rest of community comes at great cost nearing $150k/month (mostly egress) in donations.

Additionally, some of our community members are unable to access the official release artifacts due to country level firewalls that do not them connect to Google services.

Ideally we can dramatically reduce cost and allow everyone in the world to download the artifacts released by our community.

### Goals

> A policy and procedure for use by SIG Release to promote container images and release binaries to multiple registries and mirrors.

Reviewer: I think we should break this into 2 major phases:

1) Container images
2) Other artifacts

We may even want to break it to 2 KEPs so we can "finish" one.

Reviewer 2: Good thought! This KEP just focuses on container images now.

A solution to allow redirection to appropriate mirrors to lower cost and allow access from any cloud or country globally.

### Non-Goals

Anything related to creation of artifacts, bom, digital signatures.

> ## Proposal

Reviewer: We should also document that we're explicitly OK with a model where the management of the mirror is opaque to us as long as the other criteria are met.

There are two intertwined concepts that are part of this proposal.

> First, the policy and procedures to promote/upload our artifacts to multiple providers. Our existing processes upload only to GCS buckets. Ideally we extend the existing software/promotion process to push directly to multiple providers. Alternatively we use a second process to synchronize artifacts from our existing production buckets to similar constructs at other providers.

Reviewer 3: I don't think we have to go too much into technical details here, but I'd like to emphasize that modifying existing tools should be preferred over adding another step to the process. Meaning we can put the whole "Alternatively…" down to the alternatives section.

Additionally we require a registry and artifact url-redirection solution to the local cloud provider or country.

### User Stories

> #### SIG Release - Image Promotion

Reviewer 4: Do we need to ensure the compatibility with `cosign` ? (if there is a relation with [KEP 3031](https://github.com/kubernetes/enhancements/tree/master/keps/sig-release/3031-signing-release-artifacts))

Reviewer 3: cosign is compatible with OCI registries, which seems to be a trivial requirement for container registries. :+1:

```feature
As a SIG Release volunteer

> I want to promote our binaries/images to multiple clouds

Reviewer 4: ```suggestion
I want to promote our container images to multiple OCI/containers registries
```

Reviewer 4: Not needed

Given a promotion / manifest
When my PR is merged
Then the promotion process occurs
```

> #### Cloud Customer - Installing K8s via kubeadm

Reviewer 3: I'd add a use case for just pulling an official container image from region X.

```feature

> As a CLOUD end-user

Reviewer 4: We should not stick the user story to a cloud environment. We don't want to break the existing way of consuming those container images produced.

Reviewer 2: Since the most of the spend is only cloud users, this should be fine to focus on for this KEP

Author: Pulled out User Stories for this merge

I want to install Kubernetes

Given some compute resources at CLOUD
When I use kubeadm to deploy Kubernetes

> Then I will be redirected to a local CLOUD registry

Reviewer 4: we should be more clear about the meaning of "local":
```suggestion
Then I will be redirected to the closest network endpoint of the registry
```

Reviewer 2: resolved in f960d6f21b89a3d931d430e6d5e5ac18ce041107

```

### Notes/Constraints/Caveats

The primary purpose of the KEP is getting consensus on the agreed policy and procedure to unblock our community and move forward together.

There has been a lot of activity around the technology and tooling for both goals, but we need shared agreement on policy and procedure first.

### Risks and Mitigations

This is the primary pipeline for delivering Kubernetes worldwide. Ensuring the appropriate SLAs and support as well as artifact integrity is crucial.

## Design Details

> ### Artifact Promotion

Reviewer: Let's document and consider two main approaches:

1) Push.  Each mirror provider gives us a mechanism and credentials to push container images.  As we promote images from staging to prod, we push to all mirrors.  We need to consider credential security and rotation.

2) Pull.  We publish a log (git repo?) of image changes and mirrors are expected to sync changes in a reasonable period of time (99p @ 10 mins?).

Reviewer 2: Opted into a push based mechanism, where sig-k8s-infra manages the content of the buckets

#### Policy

(more details needed, #sig-release-eng?)

#### Process

Currently the promotion process is primarily driven by the CIP/[promo-tool#kpromo](https://github.com/kubernetes-sigs/promo-tools#kpromo)?

### Artifact Distribution

> #### Policy

Reviewer: We need to detail how to on-board a mirror.  E.g.:

- provide some guarantees of service (e.g. a contract with CNCF) and point of contact
- provide an emergency contact in case of outage
- provide a mapping of client IPs to mirror (maybe through a git repo)

Then we can add the mirror and have the front-end server start redirecting traffic.

We will want to periodically healthcheck each mirror (e.g. pull a random blob, measure latency).  If HC fails, remove mirror until it passes N times.  We need a site or something indicating which mirrors are healthy, maybe stats.

We will want to log all redirects and set up a PII-anonymizing process so we can publish some aggregated information about how much traffic is going to each mirror, top images globally, etc.

Reviewer 4: > provide some guarantees of service (e.g. a contract with CNCF) and point of contact

Most of the managed services for Container registries have SLAs. We need to agree about minimum level of SLA.

Reviewer 2: Going off of what @ameukam says here, we are working closely with the providers who consume the most to bring up infra that we manage.

> #### Process

Reviewer: We will need to detail how we turn up the new DNS name and redirector and how we plan to convert users of old GCR name into the new name.

Reviewer 2: Members of sig-k8s-infra have made PRs against projects like kops and Kubernetes to change the defaults.
There's also e2e testing through changing the domain for various jobs running in Prow.

## Alternatives / Background

- Apache has a widespread mirror network
  - @dims has experience here
  - http://ws.apache.org/mirrors.cgi
  - https://infra.apache.org/mirrors.html
- [Umbrella issue: k8s.gcr.io => registry.k8s.io solution k/k8s.io#1834
](https://github.com/kubernetes/k8s.io/issues/1834)
- [ii/registry.k8s.io Implementation proposals](https://github.com/ii/registry.k8s.io#registryk8sio)
- [ii.nz/blog :: Building a data pipline for displaying Kubernetes public artifact traffic
](https://ii.nz/post/building-a-data-pipline-for-displaying-kubernetes-public-artifact-traffic/)

### How much is this going to save us?

> ![Cost of K8s Artifact hosting - Data Studio Graphs](https://i.imgur.com/LAn4UIE.png)

Reviewer 4: We should provide a link to the full report and not just a screenshot for better transparency.

Reviewer 2: Update to show AWS involvement without exposing company specific usage patterns more context, in 7030358b.

## Infrastructure Needed

> It would be good to request some donations for some larger providers, including one in China, via [Cloud Native Credits program](https://www.cncf.io/credits/).

Reviewer 4: We should provide a list of what exactly is needed in terms of infrastructure to ensure this KEP go to `implementable`.

## How can we help?

When Kubernetes master opens up for v1.25 development, we need to update all default urls in our code and test harness to the new registry url. As a team sig-k8s-infra is signing up to ensure that this oci-proxy based registry.k8s.io will be as robust and available as the current setup. As a backup, we will continue to run the current k8s.gcr.io as well. So do not worry about that going away. Turning on traffic to the new url will help us monitor and fix things if/when they break and we will be able to tune traffic and lower our costs of operation.

### Goals

A policy and procedure for use by SIG Release to promote container images and release binaries to multiple registries and mirrors.

A solution to allow redirection to appropriate mirrors to lower cost and allow access from any cloud or country globally.

### Non-Goals

> Anything related to creation of artifacts, bom, digital signatures, staging buckets.

Reviewer 4: Starting 1.24, the releases will be signed. digital signatures are now in scope.

```suggestion
Anything related to creation of artifacts, bom, staging buckets.
```

Reviewer 2: Updated in f960d6f21b89a3d931d430e6d5e5ac18ce041107
